### PR TITLE
[quant][graphmode] Insert GetAttr for quantization parameters instead of Constant

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #30553 [quant][graphmode] `insert_quant_dequant` pass support shared class types
* #30860 [quant][graphmode][refactor] Factor out getInvokedMethod in `InsertQuantDeQuantHelper`
* #30859 [quant][graphmode][refactor] getQParams return a dictionary of qparams
* #30858 [quant][graphmode] Remove `values_to_quantize_`
* #30552 [quant][graphmode][refactor] Add registerQParams function
* **#30551 [quant][graphmode] Insert GetAttr for quantization parameters instead of Constant**

Summary:
To enable quantizing with shared types, we need to insert GetAttr nodes for
quantization parameters since the code might be shared by multiple module instances
and we'd like to make quantized module instance also share the same code but with
different values of attributes.

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18818652](https://our.internmc.facebook.com/intern/diff/D18818652)

### CODE REVIEW

File: torch/csrc/jit/passes/quantization.cpp [hunk 1]
```diff
@@ -637,16 +634,32 @@ void InsertQuantDeQuantHelper::removeObservers(script::Module& module) {
...
+    std::string prefix = v->debugName();
+    if (is_per_channel) {
+      module.register_attribute(prefix + "_scale", TensorType::get(), scale);
+      module.register_attribute(prefix + "_zero_point", TensorType::get(), zero_point);
+      module.register_attribute(prefix + "_axis", IntType::get(), tp->elements()[2].toInt());
+    } else {
+      module.register_attribute(prefix + "_scale", FloatType::get(), scale.item<double>());
+      module.register_attribute(prefix + "_zero_point", IntType::get(), zero_point.item<int64_t>());
+    }
+    module.register_attribute(prefix + "_scalar_type", IntType::get(), scalar_type);
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/quantization.cpp]: Could this code be moved inside `insertQuantDeQuantCall` or factored out to a separate function?

**Review by ZolotukhinM — APPROVED** [review]
Looks good, but can we have some tests for this?

**Comment by jerryzh168** [comment]
> Looks good, but can we have some tests for this?

we already have insert_quant_dequant tests, just need to make sure they still pass after the change

**Comment by ZolotukhinM** [comment]
> > Looks good, but can we have some tests for this?
> 
> we already have insert_quant_dequant tests, just need to make sure they still pass after the change

Imagine a hypothetical scenario when an engineer Bob comes up with a change that's just a revert of this PR (say, in order to simplify something). Will some test fail and stop him from doing so? IOW, if a test passed before and after the change, this change is not tested.

**Comment by jerryzh168** [comment]
This is more of a refactor of previous implementation of `insert_quant_dequant`, do we want to test for implementation details?

**Comment by ZolotukhinM** [comment]
> This is more of a refactor of previous implementation of `insert_quant_dequant`, do we want to test for implementation details?

Code refactoring is the process of restructuring existing computer code — changing the factoring —without changing its external behavior (definition from wikipedia).

Here we're clearly changing the observable behavior - what previously was `prim::Constant` now becomes `prim::GetAttr`. That's how we expect this pass to behave, so there should be tests checking that. They don't need to check exact implementation, but if it is important for some reason to generate get-attrs instead of constants there must be tests checking that in one way or another. If it was not important, then there would be no need in this PR at all.

**Comment by jerryzh168** [comment]
> > This is more of a refactor of previous implementation of `insert_quant_dequant`, do we want to test for implementation details?
> 
> Code refactoring is the process of restructuring existing computer code — changing the factoring —without changing its external behavior (definition from wikipedia).
> 
> Here we're clearly changing the observable behavior - what previously was `prim::Constant` now becomes `prim::GetAttr`. That's how we expect this pass to behave, so there should be tests checking that. They don't need to check exact implementation, but if it is important for some reason to generate get-attrs instead of constants there must be tests checking that in one way or another. If it was not important, then there would be no need in this PR at all.

I think It's an implementation details with respect to `insert_quant_dequant` function, that's why we didn't check scale/zero_point are prim::Constant when we add `insert_quant_dequant` pass either. Also I just tried and looks like we don't have a way to list all attributes in python? I tried dict but looks like it only prints the attribute that are Tensors.

**Comment by jerryzh168** [comment]
I'll land for now, we can add the tests after we support getting all attributes of a given module, which probably needs a little discussion.

**Pull request closed without merging**